enhancements/ingress/global-admission-hsts.md

```diff
@@ -0,0 +1,561 @@
+
+### Non-Goals
+
+- HSTS cannot be applied to non-TLS routes, even if HSTS is requested for all routes globally.  As
+mentioned in [RFC 6797 Section 8.1](https://tools.ietf.org/html/rfc6797#section-8.1), if an HTTP
+response is received over insecure transport, clients must ignore any present STS headers.
+
+## Proposal
+Each goal is addressed in the subsections below.
+
+### Continue to enable HSTS per-route (retain the existing feature functionality)
+The `Strict-Transport-Security` response header is currently applied per-route via an
+annotation called `haproxy.router.openshift.io/hsts_header`.  This annotation has a
+required directive `max-age`, and optional directives `includeSubDomains` and `preload`,
+with the following definitions:
+- `max-age`: delta time range in seconds during which the host is to be regarded as an HSTS host
+  - If set to 0, it negates the effect and the host is no longer regarded as an HSTS host
+  - `max-age` is a time-to-live value, so if a client makes a request for the route's host
+  and this period of time elapses before the client makes another request for the same host, the
+  HSTS policy will expire on that client
+- `includeSubDomains`: if present, applies to any subdomains of the host's domain name, when the
+route admission policy allows wildcards.  If wildcard routes aren't allowed, this directive has no
+effect.
+- `preload`: if present, tells the client to include the route's host in its host preload list so that
+it never needs to do an initial load to get the HSTS header (note that this is not defined in RFC 6797
+and is therefore client browser implementation-dependent)
+
+HSTS is currently implemented in the HAProxy template and applied to `edge` and
+`reencrypt` routes that have the `haproxy.router.openshift.io/hsts_header` annotation:
+```gotemplate
+{{- /* hsts header in response: */}}
+ {{- $hstsOptionalTokenPattern := `(?:includeSubDomains|preload)` }}
+ {{- $hstsPattern := printf `(?:%[1]s[;])*max-age=(?:\d+|"\d+")(?:[;]%[1]s)*`  $hstsOptionalTokenPattern -}}
+...
+{{- if matchValues (print $cfg.TLSTermination) "edge" "reencrypt" }}
+    {{- with $hsts := firstMatch $hstsPattern (index $cfg.Annotations "haproxy.router.openshift.io/hsts_header") }}
+  http-response set-header Strict-Transport-Security {{$hsts}}
+    {{- end }}{{/* hsts header */}}
+  {{- end }}{{/* is "edge" or "reencrypt" */}}
+```
+The HAProxy template will not change for this enhancement.  All routes must still be configured with
+the HSTS annotation.
+
+#### Provide HSTS verification per-domain
+This proposal allows cluster administrators to configure HSTS verification on a per-domain basis with
+the addition of a new openshift/api-server validating admission plugin for the router, called
+`route.openshift.io/RequiredRouteAnnotations`.  Each route must be configured with a required route
+annotation, which will be verified against the global setting on the `IngressController`.
```

Comment:
I thought in previous discussions / POCs, the required route annotations would be available via the Ingress config API (rather than in the Ingress Controller API). Has this sentiment changed?

Reply:
No, I need to fix this.